### PR TITLE
[stable8.2] Heal unencrypted file sizes at download time

### PR DESCRIPTION
**Pull request by PVince81**
Whenever the unencrypted size is -1 or equal to the encrypted size,
interpret this as an error and enable a repair routine inside the
encryption stream wrapper.

The repair routine will count how many bytes of decrypted content was
returned and then update the cache when closing the stream.

This fix is very specific to encryption but has the advantage of not impairing (slowing down) the regular download cases.

@MorrisJobke @icewind1991 @schiesbn 
- [ ] backport approval (after testing)
- [ ] add unit tests

Fixes https://github.com/owncloud/core/issues/21751 but only for encryption

### CODE REVIEW

**Comment by jknockaert** [comment]
I understand this is part of a mechanism that fixes inconsistencies in the database when they get noticed? Sounds like an excellent idea. I will have a look at it over the weekend.

**Comment by PVince81** [comment]
For every test, manually update the "size" column in the database using the real file size from disk (in the data folder)

- [ ] TEST: with small files (less than 8kb)
- [ ] TEST: with big files (more than 8kb)
- [ ] TEST: with files that have 6192 bytes (the exact unencrypted block length)
- [ ] TEST: web UI repair
- [ ] TEST: Webdav repair

**Comment by PVince81** [comment]
- [ ] TODO: we should probably also propagate the file size to the parent folders @icewind1991 ?

**Comment by MorrisJobke** [comment]
I tested this on two broken instances and it fixed the issue there. The clients are syncing again (and by this fixing the broken entries) and also the web UI works again.


- [ ] TEST: with small files (less than 8kb)
- [ ] TEST: with big files (more than 8kb)
- [ ] TEST: with files that have 6192 bytes (the exact unencrypted block length)
- [ ] TEST: web UI repair
- [ ] TEST: Webdav repair
- [ ] TEST: sync client download behavior
- [ ] TEST: smash

**Comment by MorrisJobke** [comment]
This is a :+1: from my side.

**Comment by MorrisJobke** [comment]
Just got some feedback - it works in in IE/Edge but not in Firefox or Chrome (also not on the 2nd or 3rd try). I guess we need to check this again.

**Comment by PVince81** [comment]
Hopefully not some caching issue?
Were you able to see the actual request in the network console, whether it returned the correct content-length ?

**Comment by MorrisJobke** [comment]
> Were you able to see the actual request in the network console, whether it returned the correct content-length ?

I just asked for some curl responses.

**Comment by MorrisJobke** [comment]
> Were you able to see the actual request in the network console, whether it returned the correct content-length ?

I also asked for the network console output.

**Comment by MorrisJobke** [comment]
```
Content-Length: 984680
```

curl download said:
```
961k
```

Seems to be correct.

**Comment by MorrisJobke** [comment]
@PVince81 mentioned that this could also be an off-by-one error.

**Comment by PVince81** [comment]
Ok, I was able to reproduce the bug by uploading a file that has exactly 8191 bytes.
When trying to heal it, the content-length comes out as 6126, which is wrong.

**Comment by PVince81** [comment]
Looks like my tentative optimization is wrong. For some reason the size of `$result` is bigger than unencrypted block size ?!! I'll remove the optimization and make it always use `strlen`, it's slower but safer. Considering that we only repair once, the slowness might be fine.

**Comment by PVince81** [comment]
Ok, fixed now.

I also noticed that files with smaller sizes < 6kb seem to return the correct content-length, always.
Only files close to or bigger than 8kb will have the wrong content-length returned during repair.

**Comment by MorrisJobke** [comment]
984680 is the value from the DB - but it was an unencrypted file.

**Comment by MorrisJobke** [comment]
> I also noticed that files with smaller sizes < 6kb seem to return the correct content-length, always.
> Only files close to or bigger than 8kb will have the wrong content-length returned during repair.

Let me test it.

**Comment by gig13** [comment]
@MorrisJobke Please let me know when we can communicate to the customer, they are willing to test even if it is not fully QA'd.

**Comment by MorrisJobke** [comment]
> @MorrisJobke Please let me know when we can communicate to the customer, they are willing to test even if it is not fully QA'd.

@PVince81 and I tested this locally and it worked. I think you can provide the patch for testing: https://patch-diff.githubusercontent.com/raw/owncloud/core/pull/22008.patch

Thanks @gig13

File: lib/private/files/stream/encryption.php [hunk 1]
```diff
@@ -237,6 +254,14 @@ public function stream_open($path, $mode, $options, &$opened_path) {
 			$this->readOnly = false;
 		} else {
 			$this->readOnly = true;
+
+			// in some error cases, the size from the database is the same as the encrypted size
+			if ($this->unencryptedSize < 0 || $this->unencryptedSize === $this->size) {
```
Comment (MorrisJobke) [inline lib/private/files/stream/encryption.php]: I think this is not enough. We have two instances where unencrypted size == 0 and the size is still broken. Those will never trigger this code path and therefore will never heal themselves.

Comment (PVince81) [inline lib/private/files/stream/encryption.php]: Maybe the condition could be "if $this->size is non-zero but the unencrypted size is 0", then go in repair mode.

@MorrisJobke you need to be careful where you pick your values from. Don't look at the "unencrypted_size" column in the file cache, it is unused! The unencrypted size in 8.2 is read from the "size" column.
Just as a reminder, I know it's easy to get confused.

**Comment by jknockaert** [comment]
Before commenting on the specific implementation, two questions:

1. Do we know how the database gets inconsistent? That should actually not happen.
2. Shouldn't this be part of a background proces (the cron job)? Running a query on the database to check for inconsistencies should be quite efficient.

As for the implementation I'm not sure if it is correct. When a file is opened and fseeked, the resulting fixedUnencryptedSize will still be incorrect. I would recommend an implementation where upon detecting a corrupted file size a function is called that determines the correct metrics (at one time I rewrote getFileSize to make it quite robust; not sure where it ended up), updates the wrapper variables and sets a flag such that the database gets updated upon stream_close

**Comment by PVince81** [comment]
@jknockaert 

1. so far there are some cases where the "size" column contains the encrypted size instead of the encrypted ones. In other cases the "encrypted" flag isn't set. It seems that it's due to people trying to restore encrypted files from backup and using occ files:scan afterwards. see https://github.com/owncloud/core/issues/22096

2. a background job isn't feasible because such job doesn't have access to people's passwords + encryption keys to be able to decrypt the files and find out their real sizes

File: lib/private/files/stream/encryption.php (hunk 1 above)
Comment (MorrisJobke) [inline lib/private/files/stream/encryption.php]: So this `$this->unencryptedSize` is something different than the unencrypted_size column in the DB?

**Comment by jknockaert** [comment]
The proposed fix will make things worse when it is called in the context of a http-range request: the size will be off without possibility to detect it later on; so basically the oc file system corrupts. Upon opening the encrypted file and detecting an inconsistency the correct unencrypted filesize should be determined using the getFileSize procedure I implemented in #11659 (which was never released before encryption 2.0 arrived). This implementation is efficient and robust. As for how to detect an inconsistency there are multiple ways: either a test which is evaluated upon every file opening, or maybe a flag set in a background process (but still then the flag value needs to be evaluated upon opening). In case the unencrypted filesize is updated a flag should be set to write it back to the database upon closing the stream.

File: lib/private/files/stream/encryption.php (hunk 1 above)
Comment (PVince81) [inline lib/private/files/stream/encryption.php]: Yes. `$this->unencryptedSize` is the unencrypted size, read from the "size" column in the database. `$this->size` is the encrypted size.

**Comment by PVince81** [comment]
@jknockaert I'm worried about performance or potential slow downs, that's why we're trying to do it on the fly in a single-pass. Regarding range requests, maybe in that specific situation the repairing code should simply be disabled.

Does getFileSize require re-reading the full file or is it able to somehow compute the correct size without performance impact ?

**Comment by PVince81** [comment]
- [ ] TODO: look into potential range request issue

**Comment by jknockaert** [comment]
@PVince81 getFileSize (which would have to be updated for compatibility with encryption 2.0) doesn't require re-reading the full file for seekable (encrypted) streams.
The whole encryption stream wrapper was designed under the assumption that the unencrypted file size is known upon opening the file. I would prefer a one-off performance penalty if that ensures integrity.
The corruption would happen upon every case where the file is read incompletely or an fseek is issued before closing.

**Comment by PVince81** [comment]
A one-shot penatly is acceptable, but before doing so we should detect whether a repair is needed in the first place. And only if a repair needs to be done, then do it once.

Would it work if we modified this PR to call getFileSize() + fix the cache entry at the beginning as soon as the discrepancy is detected: https://github.com/owncloud/core/pull/22008/files#diff-7e61c5b95e4e1be3406a3959ae86d2d2R262 ? Or alternatively, at the end.

**Comment by PVince81** [comment]
@jknockaert as usual, thanks a lot for your advice :smile:

**Comment by jknockaert** [comment]
@PVince81 Sounds like a good idea. So upon stream opening:
1. evaluate need to repair
2. getFileSize
3. update cache entry (and $this->unencryptedSize)
4. proceed as if nothing happened
getFileSize is no longer in the codebase; I'm not sure where and how to re-implement it (but I'd be happy to review it once it's there).

**Comment by PVince81** [comment]
@schiesbn can you comment on this ?

**Comment by jknockaert** [comment]
While I am thinking of the topic, I wonder if we should add an option somewhere to do a file scan and fix/report encryption issues. Such a scan would check for unencrypted files, incorrect file sizes, but also for the presence of all required keys. There's now far too much opportunities for corrupting the encryption process, eg. by php timeouts, without the possibility to validate. The scan would then fix what is fixable (access to private keys needed), and report everything else. I'm also thinking of the recovery keys here; the idea introduced in #18951 touches this concern. But maybe we should open a separate ticket for this.

**Comment by PVince81** [comment]
@jknockaert I thought about this too. Unfortunately occ files:scan on the CLI doesn't have the user's passwords, so unable to compute and fix unencrypted sizes. The most it could do is set the "encrypted" flag and maybe the size to "-1" to trigger this PR's logic, if "encrypted" was 0 initially.

**Comment by jknockaert** [comment]
@PVince81 What about adding this to the user settings page? There's already an encryption related section where the user can activate the recovery key functionality.

**Comment by PVince81** [comment]
@jknockaert there is a risk to run into timeouts if the PHP call takes too long.

**Comment by jknockaert** [comment]
@PVince81 OK, so the test would suffer from the same corruption that it tries to detect.

**Comment by PVince81** [comment]
So far I still think this is the best solution: https://github.com/owncloud/core/pull/22008#issuecomment-181784301

**Comment by PVince81** [comment]
@schiesbn any update on this critical issue ?

**Comment by schiesbn** [comment]
I think the easiest solution is not to try to fix the unencrypted size in fseek. Otherwise it will become to complex for the corner case we want to fix. See my last commit.

Otherwiese code from @PVince81 looks good and works as expected :+1:

**Comment by PVince81** [comment]
note for @jknockaert we decided to disable the repair routine when fseek was detected.

**Comment by PVince81** [comment]
Forward port here https://github.com/owncloud/core/pull/22567

**Comment by jknockaert** [comment]
No this is not good. The entire wrapper is designed under the assumption that the unencrypted filesize is correctly known upon access. That's why I first had to merge #11659 (to determine the unencrypted filesize without using the wrapper) before I could rewrite the wrapper in #14383. The Encryption 2.0 wrapper is conceptually identical to #14383.
Let me just give an example where the current PR goes wrong. Imagine a request for a file with a wrong size. Then an fseek happens searching from the back of the file. Because the filesize is wrong (whatever it is set to), it will result in invalid results (which may translate in file corruption).
Fixing the unencrypted file size upon opening allows to handle everything properly.

**Comment by MorrisJobke** [comment]
:+1:

**Comment by jknockaert** [comment]
@schiesbn What's complex about writing a procedure to determine unencrypted size and write it to the filecache? It would fit in nicely together in one place in the code and in a place where it is called deterministically (when opening the file). The current PR has code spread out over diverse subroutines which are called by low level php code in a fairly unpredictable manner (remember the php bug we hit in #16452).

**Comment by schiesbn** [comment]
@jknockaert after thinking once more about it I think you are right... Please have a look at this PR https://github.com/owncloud/core/pull/22579. Is this the solution you had in mind?

**Comment by PVince81** [comment]
Obsoleted by the better solution https://github.com/owncloud/core/pull/22579 which doesn't require two downloads

**Pull request closed without merging**